### ISSUE TITLE
"master not discovered yet" for 6.20+

### ISSUE THREAD

**Issue by matusfaro** [alexcojocaru/elasticsearch-maven-plugin#144]
Hello,

I was attempting to upgrade from ES `7.10` to `7.13` and found I also needed to upgrade this plugin from `6.16` to latest of `6.22`. Doing so caused an issue of `master not discovered`. I tracked down that the issue is present between `6.19` and `6.20` with [these commits](https://github.com/alexcojocaru/elasticsearch-maven-plugin/compare/elasticsearch-maven-plugin-6.19...elasticsearch-maven-plugin-6.20).

Good scenarios:
- 6.19 7.10.0

Failure scenarios:
- 6.20 7.10.0
- 6.22 7.10.0
- 6.20 7.13.4
- 6.22 7.13.4

Maven config where `elasticsearch-maven-plugin.version` is either `6.19` or `6.20`
```
            <plugin>
                <groupId>com.github.alexcojocaru</groupId>
                <artifactId>elasticsearch-maven-plugin</artifactId>
                <version>${elasticsearch-maven-plugin.version}</version>
                <configuration>
                    <version>7.10.0</version>
                    <clusterName>test</clusterName>
                    <transportPort>9300</transportPort>
                    <httpPort>9200</httpPort>
                    <skip>${skipTests}${skipITs}</skip>
                    <timeout>120</timeout>
                </configuration>
                <executions>
                    <execution>
                        <id>start-elasticsearch</id>
                        <phase>pre-integration-test</phase>
                        <goals>
                            <goal>runforked</goal>
                        </goals>
                    </execution>
                    <execution>
                        <id>stop-elasticsearch</id>
                        <phase>post-integration-test</phase>
                        <goals>
                            <goal>stop</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
```

Logs for running `mvn com.github.alexcojocaru:elasticsearch-maven-plugin:runforked` for both versions below:

`6.19` working version:
```
[INFO] --- elasticsearch-maven-plugin:6.19:runforked (default-cli) @ clearflask-server ---
[INFO] Using Elasticsearch [0] configuration: com.github.alexcojocaru.mojo.elasticsearch.v2.InstanceConfiguration@7cea0110[id=0,baseDir=/Users/matus/dev/clearflask/clearflask-server/target/elasticsearch0,httpPort=9200,transportPort=9300,pathData=<null>,pathLogs=<null>,settings=<null>]
[INFO] Elasticsearch[0]: Executing command '[bin/elasticsearch-plugin, list]' in directory '/Users/matus/dev/clearflask/clearflask-server/target/elasticsearch0'
[INFO] Elasticsearch[0]: The process finished with exit code 0
[INFO] Elasticsearch[0]: Executing command '[bin/elasticsearch, -p, pid, -Ecluster.name=test, -Ehttp.port=9200, -Etransport.tcp.port=9300, -Ehttp.cors.enabled=true]' in directory '/Users/matus/dev/clearflask/clearflask-server/target/elasticsearch0'
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
[2022-05-11T10:39:13,438][INFO ][o.e.n.Node               ] [lappy.local] version[7.10.0], pid[87943], build[oss/tar/51e9d6f22758d0374a0f3f5c6e8f3a7997850f96/2020-11-09T21:30:33.964949Z], OS[Mac OS X/12.2.1/x86_64], JVM[Amazon.com Inc./OpenJDK 64-Bit Server VM/11.0.13/11.0.13+8-LTS]
[2022-05-11T10:39:13,439][INFO ][o.e.n.Node               ] [lappy.local] JVM home [/Library/Java/JavaVirtualMachines/amazon-corretto-11.jdk/Contents/Home], using bundled JDK [false]
[2022-05-11T10:39:13,440][INFO ][o.e.n.Node               ] [lappy.local] JVM arguments [-Xshare:auto, -Des.networkaddress.cache.ttl=60, -Des.networkaddress.cache.negative.ttl=10, -XX:+AlwaysPreTouch, -Xss1m, -Djava.awt.headless=true, -Dfile.encoding=UTF-8, -Djna.nosys=true, -XX:-OmitStackTraceInFastThrow, -Dio.netty.noUnsafe=true, -Dio.netty.noKeySetOptimization=true, -Dio.netty.recycler.maxCapacityPerThread=0, -Dio.netty.allocator.numDirectArenas=0, -Dlog4j.shutdownHookEnabled=false, -Dlog4j2.disable.jmx=true, -Djava.locale.providers=SPI,COMPAT, -Xms1g, -Xmx1g, -XX:+UseConcMarkSweepGC, -XX:CMSInitiatingOccupancyFraction=75, -XX:+UseCMSInitiatingOccupancyOnly, -Djava.io.tmpdir=/var/folders/15/6r443qw917d9y61l_sw28wfc0000gn/T/elasticsearch-10127083279903283389, -XX:+HeapDumpOnOutOfMemoryError, -XX:HeapDumpPath=data, -XX:ErrorFile=logs/hs_err_pid%p.log, -Xlog:gc*,gc+age=trace,safepoint:file=logs/gc.log:utctime,pid,tags:filecount=32,filesize=64m, -XX:MaxDirectMemorySize=536870912, -Des.path.home=/Users/matus/dev/clearflask/clearflask-server/target/elasticsearch0, -Des.path.conf=/Users/matus/dev/clearflask/clearflask-server/target/elasticsearch0/config, -Des.distribution.flavor=oss, -Des.distribution.type=tar, -Des.bundled_jdk=true]
[2022-05-11T10:39:14,104][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [aggs-matrix-stats]
[2022-05-11T10:39:14,104][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [analysis-common]
[2022-05-11T10:39:14,104][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [geo]
[2022-05-11T10:39:14,104][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [ingest-common]
[2022-05-11T10:39:14,105][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [ingest-geoip]
[2022-05-11T10:39:14,105][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [ingest-user-agent]
[2022-05-11T10:39:14,105][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [kibana]
[2022-05-11T10:39:14,105][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [lang-expression]
[2022-05-11T10:39:14,105][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [lang-mustache]
[2022-05-11T10:39:14,106][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [lang-painless]
[2022-05-11T10:39:14,106][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [mapper-extras]
[2022-05-11T10:39:14,106][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [parent-join]
[2022-05-11T10:39:14,106][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [percolator]
[2022-05-11T10:39:14,107][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [rank-eval]
[2022-05-11T10:39:14,107][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [reindex]
[2022-05-11T10:39:14,107][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [repository-url]
[2022-05-11T10:39:14,107][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [transport-netty4]
[2022-05-11T10:39:14,108][INFO ][o.e.p.PluginsService     ] [lappy.local] no plugins loaded
[2022-05-11T10:39:14,141][INFO ][o.e.e.NodeEnvironment    ] [lappy.local] using [1] data paths, mounts [[/System/Volumes/Data (/dev/disk1s1)]], net usable_space [309.1gb], net total_space [1.8tb], types [apfs]
[2022-05-11T10:39:14,141][INFO ][o.e.e.NodeEnvironment    ] [lappy.local] heap size [989.8mb], compressed ordinary object pointers [true]
[2022-05-11T10:39:14,179][INFO ][o.e.n.Node               ] [lappy.local] node name [lappy.local], node ID [ET3Sovy5S5e1mmP2TT6aTQ], cluster name [test], roles [master, remote_cluster_client, data, ingest]
[2022-05-11T10:39:16,546][INFO ][o.e.t.NettyAllocator     ] [lappy.local] creating NettyAllocator with the following configs: [name=unpooled, suggested_max_allocation_size=1mb, factors={es.unsafe.use_unpooled_allocator=null, g1gc_enabled=false, g1gc_region_size=0b, heap_size=989.8mb}]
[2022-05-11T10:39:16,601][INFO ][o.e.d.DiscoveryModule    ] [lappy.local] using discovery type [zen] and seed hosts providers [settings]
[2022-05-11T10:39:16,817][WARN ][o.e.g.DanglingIndicesState] [lappy.local] gateway.auto_import_dangling_indices is disabled, dangling indices will not be automatically detected or imported and must be managed manually
[2022-05-11T10:39:16,941][INFO ][o.e.n.Node               ] [lappy.local] initialized
[2022-05-11T10:39:16,941][INFO ][o.e.n.Node               ] [lappy.local] starting ...
[2022-05-11T10:39:17,066][INFO ][o.e.t.TransportService   ] [lappy.local] publish_address {127.0.0.1:9300}, bound_addresses {[::1]:9300}, {127.0.0.1:9300}
[2022-05-11T10:39:17,357][WARN ][o.e.b.BootstrapChecks    ] [lappy.local] the default discovery settings are unsuitable for production use; at least one of [discovery.seed_hosts, discovery.seed_providers, cluster.initial_master_nodes] must be configured
[2022-05-11T10:39:17,370][INFO ][o.e.c.c.ClusterBootstrapService] [lappy.local] no discovery configuration found, will perform best-effort cluster bootstrapping after [3s] unless existing master is discovered
[2022-05-11T10:39:20,376][INFO ][o.e.c.c.Coordinator      ] [lappy.local] setting initial configuration to VotingConfiguration{ET3Sovy5S5e1mmP2TT6aTQ}
[2022-05-11T10:39:20,603][INFO ][o.e.c.s.MasterService    ] [lappy.local] elected-as-master ([1] nodes joined)[{lappy.local}{ET3Sovy5S5e1mmP2TT6aTQ}{kl7gr_6vQJSi4uyGPtilLg}{127.0.0.1}{127.0.0.1:9300}{dimr} elect leader, _BECOME_MASTER_TASK_, _FINISH_ELECTION_], term: 1, version: 1, delta: master node changed {previous [], current [{lappy.local}{ET3Sovy5S5e1mmP2TT6aTQ}{kl7gr_6vQJSi4uyGPtilLg}{127.0.0.1}{127.0.0.1:9300}{dimr}]}
[2022-05-11T10:39:20,691][INFO ][o.e.c.c.CoordinationState] [lappy.local] cluster UUID set to [89S809RaRjiY_yw8N8t-Ew]
[2022-05-11T10:39:20,764][INFO ][o.e.c.s.ClusterApplierService] [lappy.local] master node changed {previous [], current [{lappy.local}{ET3Sovy5S5e1mmP2TT6aTQ}{kl7gr_6vQJSi4uyGPtilLg}{127.0.0.1}{127.0.0.1:9300}{dimr}]}, term: 1, version: 1, reason: Publication{term=1, version=1}
[2022-05-11T10:39:20,784][INFO ][o.e.h.AbstractHttpServerTransport] [lappy.local] publish_address {127.0.0.1:9200}, bound_addresses {[::1]:9200}, {127.0.0.1:9200}
[2022-05-11T10:39:20,784][INFO ][o.e.n.Node               ] [lappy.local] started
[2022-05-11T10:39:20,891][INFO ][o.e.g.GatewayService     ] [lappy.local] recovered [0] indices into cluster_state
[INFO] The Elasticsearch instance has started
[INFO] The Elasticsearch cluster has started
```

`6.20` problematic version:
```
[INFO] --- elasticsearch-maven-plugin:6.20:runforked (default-cli) @ clearflask-server ---
[INFO] Using Elasticsearch [0] configuration: com.github.alexcojocaru.mojo.elasticsearch.v2.InstanceConfiguration@459b187a[id=0,baseDir=/Users/matus/dev/clearflask/clearflask-server/target/elasticsearch0,httpPort=9200,transportPort=9300,pathData=<null>,pathLogs=<null>,settings=<null>,startupTimeout=120]
[INFO] Elasticsearch[0]: Executing command '[bin/elasticsearch-plugin, list]' with input '' in directory '/Users/matus/dev/clearflask/clearflask-server/target/elasticsearch0'
[INFO] Elasticsearch[0]: The process finished with exit code 0
[INFO] Elasticsearch[0]: Executing command '[bin/elasticsearch, -p, pid, -Ecluster.name=test, -Ehttp.port=9200, -Etransport.tcp.port=9300, -Ediscovery.zen.ping.unicast.hosts=127.0.0.1:9300, -Ehttp.cors.enabled=true]' with input '' in directory '/Users/matus/dev/clearflask/clearflask-server/target/elasticsearch0'
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
[2022-05-11T10:40:12,031][INFO ][o.e.n.Node               ] [lappy.local] version[7.10.0], pid[88226], build[oss/tar/51e9d6f22758d0374a0f3f5c6e8f3a7997850f96/2020-11-09T21:30:33.964949Z], OS[Mac OS X/12.2.1/x86_64], JVM[Amazon.com Inc./OpenJDK 64-Bit Server VM/11.0.13/11.0.13+8-LTS]
[2022-05-11T10:40:12,032][INFO ][o.e.n.Node               ] [lappy.local] JVM home [/Library/Java/JavaVirtualMachines/amazon-corretto-11.jdk/Contents/Home], using bundled JDK [false]
[2022-05-11T10:40:12,033][INFO ][o.e.n.Node               ] [lappy.local] JVM arguments [-Xshare:auto, -Des.networkaddress.cache.ttl=60, -Des.networkaddress.cache.negative.ttl=10, -XX:+AlwaysPreTouch, -Xss1m, -Djava.awt.headless=true, -Dfile.encoding=UTF-8, -Djna.nosys=true, -XX:-OmitStackTraceInFastThrow, -Dio.netty.noUnsafe=true, -Dio.netty.noKeySetOptimization=true, -Dio.netty.recycler.maxCapacityPerThread=0, -Dio.netty.allocator.numDirectArenas=0, -Dlog4j.shutdownHookEnabled=false, -Dlog4j2.disable.jmx=true, -Djava.locale.providers=SPI,COMPAT, -Xms1g, -Xmx1g, -XX:+UseConcMarkSweepGC, -XX:CMSInitiatingOccupancyFraction=75, -XX:+UseCMSInitiatingOccupancyOnly, -Djava.io.tmpdir=/var/folders/15/6r443qw917d9y61l_sw28wfc0000gn/T/elasticsearch-14888291724868448950, -XX:+HeapDumpOnOutOfMemoryError, -XX:HeapDumpPath=data, -XX:ErrorFile=logs/hs_err_pid%p.log, -Xlog:gc*,gc+age=trace,safepoint:file=logs/gc.log:utctime,pid,tags:filecount=32,filesize=64m, -XX:MaxDirectMemorySize=536870912, -Des.path.home=/Users/matus/dev/clearflask/clearflask-server/target/elasticsearch0, -Des.path.conf=/Users/matus/dev/clearflask/clearflask-server/target/elasticsearch0/config, -Des.distribution.flavor=oss, -Des.distribution.type=tar, -Des.bundled_jdk=true]
[2022-05-11T10:40:12,716][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [aggs-matrix-stats]
[2022-05-11T10:40:12,716][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [analysis-common]
[2022-05-11T10:40:12,717][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [geo]
[2022-05-11T10:40:12,717][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [ingest-common]
[2022-05-11T10:40:12,717][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [ingest-geoip]
[2022-05-11T10:40:12,717][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [ingest-user-agent]
[2022-05-11T10:40:12,718][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [kibana]
[2022-05-11T10:40:12,718][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [lang-expression]
[2022-05-11T10:40:12,718][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [lang-mustache]
[2022-05-11T10:40:12,718][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [lang-painless]
[2022-05-11T10:40:12,718][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [mapper-extras]
[2022-05-11T10:40:12,719][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [parent-join]
[2022-05-11T10:40:12,719][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [percolator]
[2022-05-11T10:40:12,719][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [rank-eval]
[2022-05-11T10:40:12,719][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [reindex]
[2022-05-11T10:40:12,720][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [repository-url]
[2022-05-11T10:40:12,720][INFO ][o.e.p.PluginsService     ] [lappy.local] loaded module [transport-netty4]
[2022-05-11T10:40:12,720][INFO ][o.e.p.PluginsService     ] [lappy.local] no plugins loaded
[2022-05-11T10:40:12,756][INFO ][o.e.e.NodeEnvironment    ] [lappy.local] using [1] data paths, mounts [[/System/Volumes/Data (/dev/disk1s1)]], net usable_space [309.1gb], net total_space [1.8tb], types [apfs]
[2022-05-11T10:40:12,756][INFO ][o.e.e.NodeEnvironment    ] [lappy.local] heap size [989.8mb], compressed ordinary object pointers [true]
[2022-05-11T10:40:12,804][INFO ][o.e.n.Node               ] [lappy.local] node name [lappy.local], node ID [0njaIvfISpmC2UhEGXa1cQ], cluster name [test], roles [master, remote_cluster_client, data, ingest]
[2022-05-11T10:40:15,193][INFO ][o.e.t.NettyAllocator     ] [lappy.local] creating NettyAllocator with the following configs: [name=unpooled, suggested_max_allocation_size=1mb, factors={es.unsafe.use_unpooled_allocator=null, g1gc_enabled=false, g1gc_region_size=0b, heap_size=989.8mb}]
[2022-05-11T10:40:15,252][INFO ][o.e.d.DiscoveryModule    ] [lappy.local] using discovery type [zen] and seed hosts providers [settings]
[2022-05-11T10:40:15,463][WARN ][o.e.g.DanglingIndicesState] [lappy.local] gateway.auto_import_dangling_indices is disabled, dangling indices will not be automatically detected or imported and must be managed manually
[2022-05-11T10:40:15,585][INFO ][o.e.n.Node               ] [lappy.local] initialized
[2022-05-11T10:40:15,585][INFO ][o.e.n.Node               ] [lappy.local] starting ...
[2022-05-11T10:40:15,709][INFO ][o.e.t.TransportService   ] [lappy.local] publish_address {127.0.0.1:9300}, bound_addresses {[::1]:9300}, {127.0.0.1:9300}
[2022-05-11T10:40:26,029][WARN ][o.e.c.c.ClusterFormationFailureHelper] [lappy.local] master not discovered yet, this node has not previously joined a bootstrapped (v7+) cluster, and [cluster.initial_master_nodes] is empty on this node: have discovered [{lappy.local}{0njaIvfISpmC2UhEGXa1cQ}{oFHDIqNiSSeeF8fiY8G6gA}{127.0.0.1}{127.0.0.1:9300}{dimr}]; discovery will continue using [] from hosts providers and [{lappy.local}{0njaIvfISpmC2UhEGXa1cQ}{oFHDIqNiSSeeF8fiY8G6gA}{127.0.0.1}{127.0.0.1:9300}{dimr}] from last-known cluster state; node term 0, last-accepted version 0 in term 0
[2022-05-11T10:40:36,034][WARN ][o.e.c.c.ClusterFormationFailureHelper] [lappy.local] master not discovered yet, this node has not previously joined a bootstrapped (v7+) cluster, and [cluster.initial_master_nodes] is empty on this node: have discovered [{lappy.local}{0njaIvfISpmC2UhEGXa1cQ}{oFHDIqNiSSeeF8fiY8G6gA}{127.0.0.1}{127.0.0.1:9300}{dimr}]; discovery will continue using [] from hosts providers and [{lappy.local}{0njaIvfISpmC2UhEGXa1cQ}{oFHDIqNiSSeeF8fiY8G6gA}{127.0.0.1}{127.0.0.1:9300}{dimr}] from last-known cluster state; node term 0, last-accepted version 0 in term 0
[2022-05-11T10:40:46,034][WARN ][o.e.n.Node               ] [lappy.local] timed out while waiting for initial discovery state - timeout: 30s
[2022-05-11T10:40:46,036][WARN ][o.e.c.c.ClusterFormationFailureHelper] [lappy.local] master not discovered yet, this node has not previously joined a bootstrapped (v7+) cluster, and [cluster.initial_master_nodes] is empty on this node: have discovered [{lappy.local}{0njaIvfISpmC2UhEGXa1cQ}{oFHDIqNiSSeeF8fiY8G6gA}{127.0.0.1}{127.0.0.1:9300}{dimr}]; discovery will continue using [] from hosts providers and [{lappy.local}{0njaIvfISpmC2UhEGXa1cQ}{oFHDIqNiSSeeF8fiY8G6gA}{127.0.0.1}{127.0.0.1:9300}{dimr}] from last-known cluster state; node term 0, last-accepted version 0 in term 0
[2022-05-11T10:40:46,048][INFO ][o.e.h.AbstractHttpServerTransport] [lappy.local] publish_address {127.0.0.1:9200}, bound_addresses {[::1]:9200}, {127.0.0.1:9200}
[2022-05-11T10:40:46,048][INFO ][o.e.n.Node               ] [lappy.local] started
[INFO] The Elasticsearch instance has started
[INFO] The Elasticsearch cluster has started
```

**Comment by matusfaro** [comment]
As a workaround for now I use plugin version `6.19` with the `downloadUrl` feature to omit the `flavour`.

**Comment by alexcojocaru** [comment]
Thanks for the detailed bug report, Matus. That's very helpful information, I am going to look into it soon.

**Comment by alexcojocaru** [comment]
Thanks again for reporting this, Matus.
This uncovered a very old defect in the cluster status verification during startup - the post-startup tests were not checking the cluster status thoroughly and were reporting success even when the cluster was not healthy.
Once that was fixed, I found another one in the index configuration in a single-node cluster - the default replication (i.e. 1) was putting the cluster in a "yellow" state once an index was created, because shards could not be replicated to another node. I fixed it by disabling the index replication in single-node clusters.
